compiler/rustc_passes/src/dead.rs

```diff
@@ -627,7 +639,20 @@ fn check_item<'tcx>(
             }
         }
         DefKind::Impl { of_trait } => {
-            if of_trait {
+            // lints unused struct and traits after 2024, e.g.,
+            // ```rust
+            // #[derive(Debug)]
+            // struct Unused; //~ WARN
+            //
+            // trait Foo {} //~ WARN
+            // impl Foo for () {}
+            // ```
+            // but we still cannot lint unused traits whose impl blocks have methods:
+            // ```rust
+            // trait Foo { fn foo(); }
+            // impl Foo for () { fn foo() {} }
+            // ```
+            if of_trait && !tcx.hir().item(id).span.source_callsite().at_least_rust_2024() {
```

Comment:
Why gate this on an edition? This is a lint.

Reply:
Just temporary for now:)